Hadrons/DilutedNoise.hpp

```diff
@@ -332,6 +355,138 @@ bool DistillationNoise<FImpl>::mapEmpty(void) const
+    for(auto &n : noise_)
+    {
+        std::vector<unsigned char> hnoise = GridChecksum::sha256( &noise_.front().front() , sizeof(Type)*noiseSize_*noise_.size() );
+        scomb += GridChecksum::sha256_string(hnoise);
+    }
+    
+    //dilution map hash
+    std::vector<unsigned int> linmap;
+    for(auto& m : map_)
+    for(auto& p : m)
+    for(auto& t : p)
+        linmap.push_back(t);
+    std::vector<unsigned char> hmap = GridChecksum::sha256( &linmap.front() , sizeof(unsigned int)*linmap.size() );
+    
+    scomb       += GridChecksum::sha256_string(hmap);   //concatenated noise hashes and map hash
+    //hash of concatenation
+    std::vector<unsigned char> hcomb = GridChecksum::sha256( scomb.c_str() , sizeof(char)*scomb.size() ); 
+    return GridChecksum::sha256_string(hcomb);
+}
+
+template <typename FImpl>
+void DistillationNoise<FImpl>::save(const std::string filename, const std::string distilname, const unsigned int traj)
+{
+#ifdef HAVE_HDF5
+    if(grid_->IsBoss())
+    {
+        std::string filestem = filename + "." + std::to_string(traj) + ".h5";
+        Hdf5Writer writer(filestem);
+        push(writer, distilname);
+        auto &group = writer.getGroup();
+
+        //metadata write
+        //hash
+        hsize_t         hash_dim = 1;
+        std::string shash = generateHash();
+        H5NS::DataSpace hash_dataspace(1, &hash_dim);
+        H5NS::DataType  hash_datatype(H5T_STRING,shash.size());
+        H5NS::Attribute attr_hash = group.createAttribute("NoiseHash", hash_datatype, hash_dataspace);
+        attr_hash.write(hash_datatype, shash.c_str() );
+        //dilution sizes t,l,s
+        std::vector<int> dil_sizes = {dilutionSize(Index::t),dilutionSize(Index::l),dilutionSize(Index::s)};
+        hsize_t dilsize_dim = dil_sizes.size();
+        H5NS::DataSpace dil_dataspace(1, &dilsize_dim);
+        H5NS::Attribute attr_dil = group.createAttribute("DilutionSizes",  Hdf5Type<int>::type(), dil_dataspace);
+        attr_dil.write(Hdf5Type<int>::type(), dil_sizes.data());
+        //nnoise
+        unsigned int nnoise = noise_.size();
+        hsize_t nnoise_dim = 1;
+        H5NS::DataSpace nnoise_dataspace(1, &nnoise_dim);
+        H5NS::Attribute attr_nnoise = group.createAttribute("Nnoise",  Hdf5Type<unsigned int>::type(), nnoise_dataspace);
+        attr_nnoise.write(Hdf5Type<unsigned int>::type(), &nnoise);
+        
+        //noise write
+        std::vector<hsize_t>    datadim = {static_cast<hsize_t>(noiseSize_)};
+        H5NS::DataSpace         dataspace(datadim.size(), datadim.data());
+        for(unsigned int i=0 ; i<noise_.size() ; i++)
+        {
+            H5NS::DataSet           dataset = group.createDataSet("Noise_"+std::to_string(i), Hdf5Type<Type>::type(), dataspace);
+            dataset.write(&noise_[i].front() , Hdf5Type<Type>::type(), dataspace, dataspace);
+        }
+    }
+#else
+    HADRONS_ERROR(Implementation, "all-to-all matrix I/O needs HDF5 library");
```

Comment:
just a small thing, but you should write "distillation matrix I/O" here

Reply:
Done